### PR TITLE
Autoscaling & Cloudformation

### PR DESCRIPTION
**Pull request by sldsmkd**

### CODE REVIEW

File: tropopause/validators.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import re
+
+
+def valid_url(url):
```
Comment (h-bragg) [inline tropopause/validators.py:4]: worth using http://validators.readthedocs.io/en/latest/#module-validators.url ?

File: tropopause/autoscaling.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+def AddGenericUserDataRPM(func):
+    """ Add sensible cloudinit into UserData for RPM based systems """
+    def wrapper(*args, **kwargs):
+        if 'UserData' not in kwargs:
+            name = args[1]
+            # pylint: disable=line-too-long
+            cfn_signal = Join(
+                '',
+                [
+                    'INSTANCE=$(curl http://169.254.169.254/latest/meta-data/instance-id 2> /dev/null)',  # noqa
```
Comment (h-bragg) [inline tropopause/autoscaling.py]: I can see why you have done this join, but its quite hard to read what these commands are actually doing.

Comment (h-bragg) [reply tropopause/autoscaling.py]: perhaps something like:

```python
Join(
    ' && ',
    [
        'INSTANCE=$(curl http://169.254.169.254/latest/meta-data/instance-id 2> /dev/null)',
        'QUERY=AutoScalingInstances[?InstanceId==\\\'$INSTANCE\\\'].[AutoScalingGroupName]',
        'RESOURCE=$(/usr/bin/aws --region={region} autoscaling describe-auto-scaling-instances --query=$QUERY --output=text 2> /dev/null) '
            .format(region=Ref('AWS::Region')),
        '/opt/aws/bin/cfn-signal -e $? --stack {stack} --resource $RESOURCE --region {region}\n'
            .format(stack=Ref('AWS::StackName'), region=Ref('AWS::Region'))
    ]
)
```

the lines are longer, but feels easier to see what it is doing.

Comment (h-bragg) [reply tropopause/autoscaling.py]: this goes for the other uses of join too.

https://docs.python.org/3.4/library/string.html#formatspec

**Review by h-bragg — CHANGES_REQUESTED** [review]
Generally looks good. All gone a bit deep into AWS for me to spot anything implementation related.

some string formatting stuff that feels like `''.format()` will make it a whole lot easier to read.

File: tropopause/validators.py (hunk 1 above)
Comment (sldsmkd) [reply tropopause/validators.py:4]: I looked at it, this was a conscious decision not to pull in external dependencies; it also doesn't support s3://foo/bar url's if we want to use them in the future

Comment (biggianteye) [reply tropopause/validators.py:4]: Googling for "python uri validator" just shows loads of people asking how to validate urls. Meh. A gap in the market I reckon.

File: tropopause/autoscaling.py (hunk 2 above)
Comment (sldsmkd) [reply tropopause/autoscaling.py]: There's some internal cloudformation / troposphere semantics that makes it not possible to do this, i've tried the format which then embeds the internal python reference as a string, and also tried just inlining { "Ref": "AWS::Foo" }, but this gets escape slashes added.

Comment (h-bragg) [reply tropopause/autoscaling.py]: weird, it should be doing almost the same thing... you can add modifies to the fields: `{ref!s}` (i guess calls `.__str__()`) on the thing, which might help... (possibly what join is doing...)

But presumably calling `join([ 'bla ', Ref('poop'),' bla'])` should be the same as `'bla {poop} bla'.format(poop=Ref('poop'))`

**Review by h-bragg — APPROVED** [review]

**Pull request merged**